File: challenges/arrays/move zero.py
# Given an integer array nums, move all 0's to the end of it while maintaining the relative order of the non-zero elements.

# Note that you must do this in-place without making a copy of the array.

def moveZeroes(nums: list[int]) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """
    i = 0
    j = i + 1
    while i < len(nums) and j < len(nums):
        if nums[i] == 0:
            if nums[j] != 0:
                nums[i], nums[j] = nums[j], nums[i]
                i += 1
                j += 1
            else:
                j += 1
        else:
            i += 1
            j += 1 

# solution 2 snowball method
# see https://leetcode.com/problems/move-zeroes/discuss/172432/THE-EASIEST-but-UNUSUAL-snowball-JAVA-solution-BEATS-100-(O(n))-%2B-clear-explanation
#  for details
def moveZeroes2(nums: list[int]) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """
    snowball_size = 0
    for i in range(len(nums)):
        if nums[i] == 0:
            snowball_size += 1
        elif snowball_size > 0:
            nums[i - snowball_size] = nums[i]
            nums[i] = 0
            

arr = [0,1,0,3,12]
moveZeroes2(arr)
print(arr) # [1,3,12,0,0]

File: challenges/arrays/cpp/move_zero.cpp
#include <iostream>
#include <vector>
using namespace std;
class Solution
{
public:
    // complexity: O(n) but higher constant factor
    // solution 1 brute force naive
    void moveZeroes(vector<int> &nums)
    {
        int zeros = 0;
        // count the number of zeros
        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] == 0)
            {
                zeros++;
            }
        }
        // move all non-zero elements to the front
        for (int i = 0, j = 0; i < nums.size(); i++)
        {
            if (nums[i] != 0)
            {
                nums[j] = nums[i];
                j++;
            }
        }
        // fill the remaining elements with zeros
        for (int i = nums.size() - zeros; i < nums.size(); i++)
        {
            nums[i] = 0;
        }
    }

    // solution 2 two pointers
    void moveZeroes2(vector<int> &nums)
    {
        int i = 0, j = i + 1;
        for (; i < nums.size() && j < nums.size();)
        {
            if (nums[i] == 0 && nums[j] != 0)
            {
                swap(nums[i], nums[j]);
                i++;
                j++;
            }
            else if (nums[i] == 0 && nums[j] == 0)
            {
                j++;
            }
            else
            {
                i++;
                j++;
            }
        }
    }

    // solution 3 snowball approach
    void moveZeroes3(vector<int> &nums)
    {
        int snowballSize = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            if (nums[i] == 0)
            {
                snowballSize++;
            }
            else if (snowballSize > 0)
            {
                swap(nums[i], nums[i - snowballSize]);
            }
        }
    }
};

int main()
{
    Solution solution;
    vector<int> nums = {3, 0, 1, 0, 2, 0, 4};
    solution.moveZeroes2(nums);
    for (int i = 0; i < nums.size(); i++)
    {
        cout << nums[i] << " ";
    }
}